src/nxt_listen_socket.c

```diff
@@ -136,25 +206,40 @@ nxt_listen_socket_create(nxt_task_t *task, nxt_mp_t *mp,
         access = (ls->access) ? ls->access : (S_IRUSR | S_IWUSR);
 
         if (nxt_file_set_access(name, access) != NXT_OK) {
+            if (ls->start != NULL && ls->end != NULL)
+                ls->end = nxt_sprintf(ls->start, ls->end,
+                                      "nxt_file_set_access(%FN) failed",
+                                      name);
             goto listen_fail;
         }
     }
 
 #endif
 
-    nxt_debug(task, "listen(%d, %d)", s, ls->backlog);
-
-    if (listen(s, ls->backlog) != 0) {
-        nxt_alert(task, "listen(%d, %d) failed %E",
-                  s, ls->backlog, nxt_socket_errno);
-        goto listen_fail;
+    if (ls->start == NULL) {
```

Comment:
@hongzhidao How do you like this decision? It is fully consistent with the current logic of the listen and router call program.

Reply:
I think the current patch is based on the logic from the controller process, as you used `nxt_listen_socket_create()`, but it seems it requires main process does `listen` operation, and now I don't know if it's reasonable.
My suggestion is to make it clear first. Please be patient with the review.

Reply:
The moment I noted, does not perform `listen` when called `nxt_listen_socket_create` that came from the `router`, but at the same time performs it for the `controller`. That is, this logic corresponds to the already existing logic (without changes) contained in the master branch.

Reply:
1. If we perform `listen` in the `main` process before all sockets are bound, clients will be able to be accepted but without any process. It's possible to happen in the real world.
2. If we perform `listen` in the `router` process, the logic in the `nxt_listen_socket_create` is broken, because we don't check the listening connection.
```
// The below logic is required after `listen`.
#if (NXT_HAVE_UNIX_DOMAIN)

    if (orig_sa != NULL) {
        ts = nxt_socket_create(task, AF_UNIX, SOCK_STREAM, 0, 0);
        if (ts == -1) {
            goto listen_fail;
        }

        ret = connect(ts, &orig_sa->u.sockaddr, orig_sa->socklen);

        err = nxt_socket_errno;

        nxt_socket_close(task, ts);

        if (ret == 0) {
            nxt_alert(task, "connect(%d, %*s) succeed, address already in use",
                      ts, (size_t) orig_sa->length,
                      nxt_sockaddr_start(orig_sa));

            goto listen_fail;
        }
```

Reply:
**I sincerely appreciate your corrections, thank you for this!** 

It turns out that `connect` should be used after `listen` only for `controller`. And in `router` there is a connection of `socketpair`?

Reply:
I sincerely appreciate your corrections, thank you for this!

You can also move this `connect` check into a separate function like `nxt_listen_socket` and call it separately from `router` and from `nxt_listen_socket_create` with `ls->start == NULL`.

Reply:
> 1. If we perform `listen` in the `main` process before all sockets are bound, clients will be able to be accepted but without any process. It's possible to happen in the real world.
> 2. If we perform `listen` in the `router` process, the logic in the `nxt_listen_socket_create` is broken, because we don't check the listening connection.
> 
> ```
> // The below logic is required after `listen`.
> #if (NXT_HAVE_UNIX_DOMAIN)
> 
>     if (orig_sa != NULL) {
>         ts = nxt_socket_create(task, AF_UNIX, SOCK_STREAM, 0, 0);
>         if (ts == -1) {
>             goto listen_fail;
>         }
> 
>         ret = connect(ts, &orig_sa->u.sockaddr, orig_sa->socklen);
> 
>         err = nxt_socket_errno;
> 
>         nxt_socket_close(task, ts);
> 
>         if (ret == 0) {
>             nxt_alert(task, "connect(%d, %*s) succeed, address already in use",
>                       ts, (size_t) orig_sa->length,
>                       nxt_sockaddr_start(orig_sa));
> 
>             goto listen_fail;
>         }
> ```

After a little analysis of the code section you specified. It seems useless to me, because if the main task of `connect` in this code is to check that the `sockaddr` address is busy, then it is too late and not functional, which is actually confirmed by the current `EADDRINUSE` error at the `bind` stage. What do you think about this?

Reply:
Note that it connects the original sockaddr, not the temp one.
```
ret = connect(ts, &orig_sa->u.sockaddr, orig_sa->socklen);
```

Reply:
That's right, I forgot that EADDRINUSE occurs in `nxt_main_listening_socket` which does not have an unlink. Yes, here, under this condition, it looks legitimate. I'll think about how to solve this now. Thank you.

Reply:
@hongzhidao Good morning. 

Please take a look at the latest commit. I tried to implement all your wishes regarding `listen` in `router`, but also for `controller` and the need to check `connect` and then `rename`.

Reply:
@hongzhidao Good morning. 

Please take a look at the latest commit: https://github.com/nginx/unit/pull/670/commits/032e63afe9a504395f893a8b509dc7a073380ab9. I tried to implement all your wishes regarding `listen` in `router`, but also for `controller` and the need to check `connect` and then `rename`.

Reply:
Sorry that personally, I can't agree this new way.

1. It seems the logic is broken.
```
static void
nxt_router_conf_error(nxt_task_t *task, nxt_router_temp_conf_t *tmcf)
{
    ...
    for (qlk = nxt_queue_first(&creating_sockets);
         qlk != nxt_queue_tail(&creating_sockets);
         qlk = nxt_queue_next(qlk))
    {
        skcf = nxt_queue_link_data(qlk, nxt_socket_conf_t, link);
        s = skcf->listen->socket;

        if (s != -1) {
            nxt_socket_close(task, s);
        }

        nxt_free(skcf->listen);
    }
```

2. It's strange that the `router` process has to maintain one more socket information about `rename`. I think it makes things complicated.

We think the ideal solution is to clean up all the unused unix domain socket files.
This is our internal plan for the future.

Reply:
Thanks for your correction.

I'll fix `nxt_router_conf_error`, thanks for bringing this to my attention.

But `rename` cannot be performed directly from `router`, as it does not have sufficient permissions to perform this operation. This moment is tested by me personally.

Please tell me more about your ideal solution. I want to note that cleaning up unused (abandoned) sockets and the problem of reusing sockets at the moment are completely different problems.

The whole difficulty is that you want to ensure similar work of `controller` and `router` on the basis of one `nxt_listen_socket_create`, depending on the place in the code where `listen` is called, and also provide all the necessary checks and operations after running `listen` ( `connect`, `rename`).

Reply:
> Please tell me more about your ideal solution. I want to note that cleaning up unused (abandoned) sockets and the problem of reusing sockets at the moment are completely different problems.

If we can clean up unused sockets, then we don't need the way borrowed from `controller`, though it's more robust.

> But rename cannot be performed directly from router, as it does not have sufficient permissions to perform this operation. This moment is tested by me personally.

Right, I agree that the way works, but it makes the router more complicated. The router code is already complicated, I prefer not to make it bigger...

Reply:
The problem is that abandoned and deleted `router` sockets during a successful (soft) termination may not solve the problem of reusing sockets when the program terminates unexpectedly.

Removing existing sockets must somehow be done during the socket creation process, as it may have been abandoned at the time of the previous unexpected termination of the program.

Previously (https://github.com/nginx/unit/pull/670#discussion_r851778433) I suggested not using `nxt_listen_socket_create` and adding `unlink` before `bind` in `nxt_main_listening_socket`, but you (https://github. com/nginx/unit/pull/670#discussion_r851779942) said that it should be abandoned in favor of `nxt_listen_socket_create`, which involves communication with `controller` and the need to separate the logic of `listen`, `connect`, `rename`.

Reply:
> Sorry that personally, I can't agree this new way.
> 
> 1. It seems the logic is broken.

Seems to be fixed now:
https://github.com/nginx/unit/pull/670/commits/e1cea5d7dd10eb0ec27f39be3c595dbd66a141fa